elm/wizard.py

```diff
@@ -603,12 +619,62 @@ def query_vector_db(self, query, probes=25, limit=100):
 
         return strings, scores, best
 
+    def _format_refs(self, refs, ids):
+        """Parse and nicely format a reference dictionary into a list of well
+        formatted string representations
+
+        Parameters
+        ----------
+        refs : list
+            List of references returned from the vector db
+        ids : np.ndarray
+            IDs of the used text from the text corpus sorted by embedding
+            relevance.
+
+        Returns
+        -------
+        out : list
+            Unique ordered list of references (most relevant first)
+        """
+
+        ref_list = []
+        for item in refs:
+            ref_dict = {}
+            for icol, col in enumerate(self.meta_columns):
+                value = item[icol]
+                value = str(value).replace(chr(34), '')
+                ref_dict[col] = value
+
+            ref_list.append(ref_dict)
+
+        seen = set()
+        unique_ref_list = []
+        for ref_dict in ref_list:
+            if str(ref_dict) not in seen:
+                seen.add(str(ref_dict))
+                unique_ref_list.append(ref_dict)
+        ref_list = unique_ref_list
+
+        if 'id' in ref_list[0]:
+            ids_list = list(ids)
+            sorted_ref_list = []
+            for ref_id in ids_list:
+                for ref_dict in ref_list:
+                    if ref_dict['id'] == ref_id:
+                        sorted_ref_list.append(ref_dict)
+                        break
+            ref_list = sorted_ref_list
```

Comment:
So it took me several min to trace this code:
```
ids_list = list(ids)
sorted_ref_list = []
for ref_id in ids_list:
    for ref_dict in ref_list:
        if ref_dict['id'] == ref_id:
            sorted_ref_list.append(ref_dict)
            break
ref_list = sorted_ref_list
```

You have a double `for` loop with an `if` clause at the end, so there is a lot to track in order to understand what you are trying to do here. It's fine - it works and I was eventually able to understand what you are getting at. 

Personally, however, I would opt to break down the for loops a little bit and try make the code a little more readable, in my option. Maybe something like this:

```
def has_relevance_score(ref_dict):
    return ref_dict.get("id") in ids

def embedding_relevance(ref_dict):
    return np.argwhere(ids == ref_dict.get("id"))

ref_list = filter(has_relevance_score, ref_list)
ref_list = sorted(ref_list, key=embedding_relevance)
```

This is just my 2 cents, so I'll leave it up to you what version you want to stick with

Reply:
I may leave this as is for now and look into it a bit more.